src/Microsoft.AspNetCore.Blazor.Browser/Interop/JavaScriptInvoke.cs

```diff
@@ -0,0 +1,172 @@
+            }
+        }
+
+        private static object InvokeDotNetMethodCore(string registration, string callbackId, string methodArguments)
+        {
+            if (!ResolvedFunctions.TryGetValue(registration, out var registeredFunction))
+            {
+                throw new InvalidOperationException($"No method exists with registration number '{registration}'.");
+            }
+
+            if (!(registeredFunction is Func<string, object> invoker))
+            {
+                throw new InvalidOperationException($"The registered invoker has the wrong signature.");
+            }
+
+            var result = invoker(methodArguments);
+            if (callbackId != null && !(result is Task))
+            {
+                var methodSpec = ResolvedFunctionRegistrations.Single(kvp => kvp.Value == registration);
+                var options = JsonUtil.Deserialize<MethodOptions>(methodSpec.Key);
+                throw new InvalidOperationException($"'{options.Method.Name}' in '{options.Type.Name}' must return a Task.");
+            }
+
+            if (result is Task && callbackId == null)
+            {
+                var methodSpec = ResolvedFunctionRegistrations.Single(kvp => kvp.Value == registration);
+                var options = JsonUtil.Deserialize<MethodOptions>(methodSpec.Key);
+                throw new InvalidOperationException($"'{options.Method.Name}' in '{options.Type.Name}' must not return a Task.");
+            }
+
+            if (result is Task taskResult)
+            {
+                // For async work, we just setup the callback on the returned task to invoke the appropiate callback in JavaScript.
+                SetupResultCallback(callbackId, taskResult);
+
+                // We just return null here as the proper result will be returned through invoking a JavaScript callback when the
+                // task completes.
+                return null;
+            }
+            else
+            {
+                return result;
+            }
+        }
+
+        private static void SetupResultCallback(string callbackId, Task taskResult)
+        {
+            taskResult.ContinueWith(task =>
+            {
+                if (task.Status == TaskStatus.RanToCompletion)
+                {
+                    if (task.GetType() == typeof(Task))
+                    {
+                        RegisteredFunction.Invoke<bool>(
+                            InvokePromiseCallback,
+                            callbackId,
+                            new InvocationResult<object> { Succeeded = true, Result = null });
+                    }
+                    else
+                    {
+                        var resultProperty = task.GetType()
+                                .GetProperties(BindingFlags.Public | BindingFlags.Instance | BindingFlags.NonPublic)
+                                .SingleOrDefault(m => m.Name == "ResultOnSuccess") ??
```

Comment:
From the other conversation:

> We can't avoid reflection here I think, given that we need to get it through ResultOnSuccess which is a mono specific element. Another option will probably to build a delegate for a given Task to access the result.

Could you explain why we need to use `ResultOnSuccess`?

I just tried using [this utility code](https://gist.github.com/SteveSandersonMS/c3a205f19264486753183f5b75c5d6b5) to obtain the `Task` result on Mono WebAssembly and it worked fine. Is that definitely not sufficient for our needs here?

Reply:
This didn't work when I tried it out, maybe with an updated mono version it started working. I'm giving your helper a try